File: python/mirage/resource/databricks_volume/databricks_volume.py
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

import dataclasses
from typing import Any

from mirage.accessor.databricks_volume import DatabricksVolumeAccessor
from mirage.commands.builtin.databricks_volume import \
    COMMANDS as DATABRICKS_VOLUME_COMMANDS
from mirage.core.databricks_volume.exists import exists
from mirage.core.databricks_volume.glob import resolve_glob as _resolve_glob
from mirage.core.databricks_volume.read import read_bytes
from mirage.core.databricks_volume.readdir import readdir
from mirage.core.databricks_volume.stat import stat as databricks_stat
from mirage.core.databricks_volume.stream import range_read, read_stream
from mirage.ops.databricks_volume import OPS as DATABRICKS_VOLUME_OPS
from mirage.resource.base import BaseResource
from mirage.resource.databricks_volume.config import DatabricksVolumeConfig
from mirage.resource.databricks_volume.prompt import PROMPT
from mirage.types import PathSpec, ResourceName

_DATABRICKS_VOLUME_OPS = {
    "read_bytes": read_bytes,
    "readdir": readdir,
    "stat": databricks_stat,
    "read_stream": read_stream,
    "range_read": range_read,
    "exists": exists,
}


class DatabricksVolumeResource(BaseResource):
    name: str = ResourceName.DATABRICKS_VOLUME
    is_remote: bool = True
    _ops: dict[str, Any] = _DATABRICKS_VOLUME_OPS
    PROMPT: str = PROMPT

    def __init__(
        self,
        config: DatabricksVolumeConfig,
        client: Any | None = None,
    ) -> None:
        super().__init__()
        self.config = config
        self.accessor = DatabricksVolumeAccessor(self.config, client)

        for fn in DATABRICKS_VOLUME_COMMANDS:
            self.register(fn)
        for fn in DATABRICKS_VOLUME_OPS:
            self.register_op(fn)

    async def resolve_glob(self, paths, prefix: str = ""):
        if prefix:
            paths = [
                dataclasses.replace(p, prefix=prefix)
                if isinstance(p, PathSpec) and not p.prefix else p
                for p in paths
            ]
        return await _resolve_glob(self.accessor, paths, self._index)

    def get_state(self) -> dict:
        redacted = ["token"]
        cfg = self.config.model_dump()
        for field in redacted:
            if cfg.get(field) is not None:
                cfg[field] = "<REDACTED>"
        return {
            "type": self.name,
            "needs_override": True,
            "redacted_fields": redacted,
            "config": cfg,
        }

    def load_state(self, state: dict) -> None:
        pass

File: typescript/packages/node/src/resource/databricks_volume/databricks_volume.ts
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

import {
  BaseResource,
  DATABRICKS_VOLUME_COMMANDS,
  DATABRICKS_VOLUME_OPS,
  DATABRICKS_VOLUME_PROMPT,
  DatabricksVolumeAccessor,
  databricksVolumeCopy,
  databricksVolumeCreate,
  databricksVolumeExists,
  databricksVolumeFind,
  databricksVolumeMkdir,
  databricksVolumeRangeRead,
  databricksVolumeRead,
  databricksVolumeReadStream,
  databricksVolumeReaddir,
  databricksVolumeRename,
  databricksVolumeRmRecursive,
  databricksVolumeRmdir,
  databricksVolumeStat,
  databricksVolumeUnlink,
  databricksVolumeWrite,
  type FileStat,
  type FindOptions,
  PathSpec,
  type RegisteredCommand,
  type RegisteredOp,
  type Resource,
  ResourceName,
  resolveDatabricksVolumeGlob,
} from '@struktoai/mirage-core'
import {
  redactDatabricksVolumeConfig,
  type DatabricksVolumeConfig,
  type DatabricksVolumeConfigRedacted,
} from './config.ts'
import { loadDatabricksProfile } from './profile.ts'

export interface DatabricksVolumeResourceState {
  type: string
  config: DatabricksVolumeConfigRedacted
}

async function resolveAuth(config: DatabricksVolumeConfig): Promise<[string, string]> {
  let host = config.host ?? process.env.DATABRICKS_HOST
  let token = config.token ?? process.env.DATABRICKS_TOKEN
  if (host === undefined || host === '' || token === undefined || token === '') {
    const profileName = config.profile ?? process.env.DATABRICKS_CONFIG_PROFILE ?? 'DEFAULT'
    const profile = await loadDatabricksProfile(profileName)
    host = host !== undefined && host !== '' ? host : profile.host
    token = token !== undefined && token !== '' ? token : profile.token
  }
  if (host === undefined || host === '' || token === undefined || token === '') {
    throw new Error(
      'databricks_volume: missing credentials; set host/token in the config, ' +
        'DATABRICKS_HOST/DATABRICKS_TOKEN env vars, or a ~/.databrickscfg profile',
    )
  }
  return [host, token]
}

export class DatabricksVolumeResource extends BaseResource implements Resource {
  readonly kind: string = ResourceName.DATABRICKS_VOLUME
  readonly isRemote: boolean = true
  readonly indexTtl: number = 600
  readonly prompt: string = DATABRICKS_VOLUME_PROMPT
  readonly config: DatabricksVolumeConfig
  readonly accessor: DatabricksVolumeAccessor
  readonly opsMap: Record<string, unknown> = {
    read_bytes: databricksVolumeRead,
    write: databricksVolumeWrite,
    readdir: databricksVolumeReaddir,
    stat: databricksVolumeStat,
    read_stream: databricksVolumeReadStream,
    range_read: databricksVolumeRangeRead,
    exists: databricksVolumeExists,
    create: databricksVolumeCreate,
    unlink: databricksVolumeUnlink,
    mkdir: databricksVolumeMkdir,
    rmdir: databricksVolumeRmdir,
    copy: databricksVolumeCopy,
    rename: databricksVolumeRename,
    rm_recursive: databricksVolumeRmRecursive,
  }

  private constructor(config: DatabricksVolumeConfig, accessor: DatabricksVolumeAccessor) {
    super()
    this.config = config
    this.accessor = accessor
  }

  static async create(config: DatabricksVolumeConfig): Promise<DatabricksVolumeResource> {
    const [host, token] = await resolveAuth(config)
    const accessor = new DatabricksVolumeAccessor(config, host, token)
    return new DatabricksVolumeResource(config, accessor)
  }

  open(): Promise<void> {
    return Promise.resolve()
  }

  close(): Promise<void> {
    return Promise.resolve()
  }

  commands(): readonly RegisteredCommand[] {
    return DATABRICKS_VOLUME_COMMANDS
  }

  ops(): readonly RegisteredOp[] {
    return DATABRICKS_VOLUME_OPS
  }

  streamPath(p: PathSpec): AsyncIterable<Uint8Array> {
    return databricksVolumeReadStream(this.accessor, p)
  }

  readFile(p: PathSpec): Promise<Uint8Array> {
    return databricksVolumeRead(this.accessor, p)
  }

  writeFile(p: PathSpec, data: Uint8Array): Promise<void> {
    return databricksVolumeWrite(this.accessor, p, data)
  }

  async appendFile(p: PathSpec, data: Uint8Array): Promise<void> {
    let existing: Uint8Array
    try {
      existing = await databricksVolumeRead(this.accessor, p)
    } catch (err) {
      if ((err as { code?: string } | null)?.code === 'ENOENT') {
        existing = new Uint8Array()
      } else {
        throw err
      }
    }
    const merged = new Uint8Array(existing.byteLength + data.byteLength)
    merged.set(existing, 0)
    merged.set(data, existing.byteLength)
    await databricksVolumeWrite(this.accessor, p, merged)
  }

  readdir(p: PathSpec): Promise<string[]> {
    return databricksVolumeReaddir(this.accessor, p, this.index)
  }

  stat(p: PathSpec): Promise<FileStat> {
    return databricksVolumeStat(this.accessor, p)
  }

  exists(p: PathSpec): Promise<boolean> {
    return databricksVolumeExists(this.accessor, p)
  }

  mkdir(p: PathSpec): Promise<void> {
    return databricksVolumeMkdir(this.accessor, p, undefined, true)
  }

  rmdir(p: PathSpec): Promise<void> {
    return databricksVolumeRmdir(this.accessor, p)
  }

  unlink(p: PathSpec): Promise<void> {
    return databricksVolumeUnlink(this.accessor, p)
  }

  rename(src: PathSpec, dst: PathSpec): Promise<void> {
    return databricksVolumeRename(this.accessor, src, dst)
  }

  copy(src: PathSpec, dst: PathSpec): Promise<void> {
    return databricksVolumeCopy(this.accessor, src, dst)
  }

  async rmR(p: PathSpec): Promise<void> {
    await databricksVolumeRmRecursive(this.accessor, p)
  }

  find(p: PathSpec, options: FindOptions = {}): Promise<string[]> {
    return databricksVolumeFind(this.accessor, p, options, this.index)
  }

  glob(paths: readonly PathSpec[], prefix = ''): Promise<PathSpec[]> {
    const effective = prefix
      ? paths.map((p) =>
          p.prefix
            ? p
            : new PathSpec({
                original: p.original,
                directory: p.directory,
                ...(p.pattern !== null ? { pattern: p.pattern } : {}),
                resolved: p.resolved,
                prefix,
              }),
        )
      : paths
    return resolveDatabricksVolumeGlob(this.accessor, effective, this.index)
  }

  getState(): Promise<DatabricksVolumeResourceState> {
    return Promise.resolve({
      type: this.kind,
      config: redactDatabricksVolumeConfig(this.config),
    })
  }

  loadState(_state: DatabricksVolumeResourceState): Promise<void> {
    return Promise.resolve()
  }
}
